ath79/ar71xx: Complete support for TP-Link CPE210 v2/v3

adschm opened the pull request:
At the moment, the CPE210 v2 is supported on ar71xx.
Support for the CPE210 v3 is prepared in a merge-ready PR (https://github.com/openwrt/openwrt/pull/1843).

This PR aims at completing the support for those devices by:
- Adding support for the CPE210 v2 also to ath79 (which requires dynamic partitioning on ar71xx)
- Adding support for the CPE210 v3 also to ar71xx (based on the old PR https://github.com/openwrt/openwrt/pull/1279)

Note that the first patch is just a copy/paste from https://github.com/openwrt/openwrt/pull/1843, because it is a required dependency. Please comment this patch there.

Both devices are tested with both targets, factory and sysupgrade.

robimarko committed:
ar71xx: Add support for TP-Link CPE210 v3

Looks identical to the v2.

This PR adds support for a popular low-cost 2.4GHz N based AP

Specifications:
 - SoC: Qualcomm Atheros QCA9533 (650MHz)
 - RAM: 64MB
 - Storage: 8 MB SPI NOR
 - Wireless: 2.4GHz N based built into SoC 2x2
 - Ethernet: 1x 100/10 Mbps, integrated into SoC, 24V POE IN

Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

Signed-off-by: Robert Marko <robimarko@gmail.com>
[Rebased, adjusted for separate tplink-safeloader entry and
dynamic partitioning, added SUPPORTED_DEVICES for ath79]
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adschm committed:
ath79: Add support for TP-Link CPE210 v2

This PR adds support for a popular low-cost 2.4GHz N based AP

Specifications:
 - SoC: Qualcomm Atheros QCA9533 (650MHz)
 - RAM: 64MB
 - Storage: 8 MB SPI NOR
 - Wireless: 2.4GHz N based built into SoC 2x2
 - Ethernet: 1x 100/10 Mbps, integrated into SoC, 24V POE IN

Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on
for around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP:192.168.0.100
Stock device TFTP adress:192.168.0.254

This is based on the support patch for the identical CPE210 v3
by Mario Schroen <m.schroen@web.de>.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adschm committed:
ar71xx: Use dynamic partitions for TP-Link CPE210 v2

This is also helpful to add support in ath79.

Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>
Tested-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

bluelineXY commented:
Hi, 
if i compare the dts files of the v2 and v3 there is only the naming of the device different. 

Would it make sense to build a generic dts file for both and only containing the name (like it's done for qca9558_tplink_tl-wr1043nd-v3.dts)

If you agrre to my proposal in the other PR i can take this part.

adrianschmutzler commented:
I thought about that, too.

There is a similar discussion going on here:
https://github.com/openwrt/openwrt/pull/1748#issuecomment-477407881

As pointed out there, I hesitate to create a merged dtsi now for two devices. I would rather wait until we have more devices supported and then try to find a resonable dtsi scheme which covers the majority of CPExxx devices.
This will also reduce mutual dependencies for the many PRs around at the moment.

In addition, since everything is at least basically reasonable and tested at the moment, I would stop moving stuff around and just keep it that way until we receive feedback from the committers.

apcameron commented:
@adrianschmutzler In order to have the status lights consistent please consider adding the aliases section to the dts files
EG
  	aliases { 
  		led-boot = &system; 
   		led-failsafe = &system; 
  		led-running = &system; 
  		led-upgrade = &system; 
  	}; 

Refer to PR#1704 and PR#1748 as examples (In PR#1738 I do not have the led-running as there are not enough leds)

adrianschmutzler commented:
@apcameron What's the common practice on this? You are using link4 if I'm not mistaken ...?
Power LED is not accessible?

apcameron commented:
Yes. Iink4 is what I have seen used for the CPE devices except for the CPE610 which only has 3 leds in total. As far as the Power LED I do not think we can use it.

adrianschmutzler commented:
@apcameron Thanks for the advice, I implemented that. Actually, I was missing a status LED before :-)

adrianschmutzler commented:
Updated and retested successfully

bluelineXY committed:
ath79: add TP-Link CPE210 v3

Specifications:

    * SoC: Qualcomm Atheros QCA9533 (650MHz)
    * RAM: 64MB
    * Storage: 8 MB SPI NOR
    * Wireless: 2.4GHz N based built into SoC 2x2
    * Ethernet: 1x 100/10 Mbps, integrated into SoC, 24V POE IN

Installation:
    Flash factory image through stock firmware WEB UI or TFTP
    To get to TFTP recovery just hold reset button while powering
    on for around 4-5 seconds and release.
    Rename factory image to recovery.bin
    Stock TFTP server IP:192.168.0.100
    Stock device TFTP adress:192.168.0.254

Thanks to robimarko for the work inside the ar71xx tree.
Thanks to adrianschmutzler for deep discussion and fixes.

Signed-off-by: Mario Schroen <m.schroen@web.de>
[Split into DTS/DTSI]
Signed-off-by: Adrian Schmutzler <freifunk@adrianschmutzler.de>

adrianschmutzler commented:
Updated ath79 v3 patch from source PR (still all tested).

ynezz commented:
I'm wondering what's the reason for not having the common `qca9533_tplink_cpe210.dtsi` DTSI?

```diff
--- target/linux/ath79/dts/qca9533_tplink_cpe210-v2.dts	2019-04-01 16:00:34.733097423 +0200
+++ target/linux/ath79/dts/qca9533_tplink_cpe210-v3.dts	2019-04-01 15:59:15.361094158 +0200
@@ -7,8 +7,8 @@
 #include "qca953x.dtsi"
 
 / {
-	compatible = "tplink,cpe210-v2", "qca,qca9533";
-	model = "TP-LINK CPE210 v2";
+	compatible = "tplink,cpe210-v3", "qca,qca9533";
+	model = "TP-LINK CPE210 v3";
 
 	chosen {
 		bootargs = "console=ttyS0,115200n8";
@@ -125,4 +125,3 @@
 	mtd-cal-data = <&ART 0x1000>;
 	mtd-mac-address = <&info 0x8>;
 };
-
```

ynezz commented on tools/firmware-utils/src/tplink-safeloader.c:
```diff
@@ -182,8 +182,7 @@ static struct device_info boards[] = {
 			{"device-id", 0x33000, 0x00100},
-			{"os-image", 0x40000, 0x1c0000},
-			{"file-system", 0x200000, 0x5b0000},
+			{"firmware", 0x40000, 0x770000},
```
@adrianschmutzler Aren't you supposed to change `first_sysupgrade_partition` to `firmware` as well?

https://github.com/openwrt/openwrt/blob/6b2d917d0c09ad6ce594b0836291335febe60a35/tools/firmware-utils/src/tplink-safeloader.c#L195

adrianschmutzler replied:
No, I made the same error for the CPE210v3 PR; it seems to have to be the name of dynamically created partition (all other devices also do it this way):
https://github.com/openwrt/openwrt/pull/1843#discussion_r268333861

ynezz replied:
Ok, seems like nobody has done this change so far, weird. Anyway, any reason to make `{"firmware", 0x40000, 0x770000},` shorter by 256k ? I mean `0x40000+0x1c0000+0x5b0000 = 0x7b0000` and not `0x770000` ?

adrianschmutzler replied:
0x770000 is the _length_?! Or am I getting something terribly wrong?

0x1c0000+0x5b0000 =0x770000 ??

adrianschmutzler replied:
> Ok, seems like nobody has done this change so far, weird. 

From what I get out of @bluelineXY 's report, using "firmware" instead of os-image will not work:
https://github.com/openwrt/openwrt/pull/1843#discussion_r268333861

ynezz replied:
Ok, it's base + size, so `0x40000+0x770000=0x7b0000`

adrianschmutzler commented:
> I'm wondering what's the reason for not having the common qca9533_tplink_cpe210.dtsi DTSI?

CPE210v1 (and CPE510v1) are two-port devices. CPE210v2 and v3 are one-port devices. Since there are a lot of other CPExxx devices waiting to be merged, my intention was to wait whether we find a better solution for all CPE devices after some of them have been merged. Otherwise, we might end up with 4 nested levels of dtsis, which in my opinion is not desirable.
However, if you still consider a dtsi for these TWO devices desirable, I will change this.

ynezz commented:
> However, if you still consider a dtsi for these TWO devices desirable, I will change this.

Yes, please. I prefer to have one DTSI instead of 2*120 copy&pasted identical lines for "just in case". While at it please also:

* remove the stray new line at the end of the v2 DTS file.
* move `SUPPORTED_DEVICES += cpe210-v3` from ar71xx to ath79

Otherwise it looks good, and I'm about to merge it. Good work everyone!

adrianschmutzler commented:
runtested on both devices

ynezz commented:
Thanks, I've merged it with few small tweaks into my staging tree https://git.openwrt.org/openwrt/staging/ynezz.git

ynezz commented:
Please take a look again and let me know if it's ok.

adrianschmutzler commented:
It's ok.

I would move the
`&eth1 { compatible = "syscon", "simple-mfd"; };`
to the individual dts files, but can live with you not touching it anymore.

Note that technically the partition layout for the v2 in ar71xx has to be changed before the v2 is added in ath79, but it will hardly matter as long as it is not merged separately.

Thanks for merging this!

ynezz commented:
> and thus won't be relevant for a CPE210v1

Thanks, but I've fallen into this trap few times already, so let's do this refactoring once there is some real user, who can actually test, that it could work with this simple changes. Until then I'll simply leave it as it is.

> Note that technically the partition layout for the v2 in ar71xx has to be changed before the v2 is added in ath79, but it will hardly matter as long as it is not merged separately.

If I understand it correctly, current v2 user on ar71xx with fixed partition layout would first need to upgrade to ar71xx with this new dynamic partition layout and then could sysupgrade to ath79. If user doesn't follow this order, then the user will get device in unbootable state, but still could flash the latest release via TFTP recovery. In other words, we're not going to brick the device.

Anyway, it would be nice if this could be added to the wiki page of this device.

adrianschmutzler commented:
Okay, everything fine!

adrianschmutzler commented on target/linux/ath79/dts/qca9533_tplink_cpe210-v2.dtsi:
```diff
@@ -0,0 +1,124 @@
+
+			config: partition@7c0000 {
+				label = "config";
+				reg = <0x7c0000 0x30000>;
```
Here is a "read-only;" missing.

adrianschmutzler commented:
@ynezz Just realized that a "read-only;" is missing for the config partition in the DTSI.
Marked it as a review above. Either change it or I will send a patch after merge, since it is not technically destroying anything.

ynezz commented:
@adrianschmutzler https://git.openwrt.org/?p=openwrt/staging/ynezz.git;a=commit;h=987a0b7a48872e5c363590ee71172d3e37f31bef

adrianschmutzler commented:
LGTM (only checked readonly)

adrianschmutzler commented:
WiFi seems to be broken on both v2/v3: https://bugs.openwrt.org/index.php?do=details&task_id=2333

adrianschmutzler commented:
@bluelineXY That's very interesting. I've always wondered how I could have overlooked a defective WiFi when adding support for those devices. But if really v3 is working and v2 is not, I can at least explain that.

Still, I have no clue how this can be caused on CPE210v2 (and CPE220V3 according to @apcameron).
Will read through the forum thread when I have some more time.

bluelineXY commented:
Yup. 
i double checked with our comited version and it works out of the box ... 
then i did a sysupgrade to latest git version from git.openwrt.org also working. 
Was able to create an AP and connect  - on both versions. 

the only tricky thing was i had to do 

ifconfig wlan0 up 
and set the disabled in the wireless config to 0


I ask him to do so ...

apcameron commented:
@bluelineXY  it is not working on the Image created by the buildbot and being downloaded by everyone.
Please download https://downloads.openwrt.org/snapshots/targets/ath79/generic/openwrt-ath79-generic-tplink_cpe210-v2-squashfs-sysupgrade.bin and test it again.

apcameron commented:
I did what you asked and it still does not work
ifconfig wlan0 up
and set the disabled in the wireless config to 0

bluelineXY commented:
@adrianschmutzler  would it be possible that you test the v2 ? v3 works for me even with the buildbot image. 
Some output from my tests are here: 
https://forum.openwrt.org/t/support-for-tp-link-cpe210-v3-0/20208/44

adrianschmutzler commented:
I can have another look during the weekend ... (However, so far it has never worked).

bluelineXY commented:
you mean ever worked, right ?

apcameron commented:
@adrianschmutzler Have you had a chance to test the wifi on CPE210 V2 using the buildbot image at https://downloads.openwrt.org/snapshots/targets/ath79/generic/openwrt-ath79-generic-tplink_cpe210-v2-squashfs-sysupgrade.bin  
I think most folks have loaded it using sysupgrade and not from the factory image (Not sure if that will make any difference)

adrianschmutzler commented:
@apcameron I've tried with a custom-built one and it's definitely broken on v2. I can also retry with official snapshot. I do not see any reason why factory and sysupgrade should differ for that.

apcameron commented:
@adrianschmutzler @bluelineXY @ynezz @chunkeey Any Ideas on who we can ask to determine the root cause for the Wifi Not working for some people?
I have seen a number of people reporting this issue so we need to resolve it soon.
Could it be related to the region the device was purchased in? (But from what I have heard it is failing for everyone on the TP-LINK CPE210 V2)

bluelineXY commented:
Hi, 
the interesting point is it even fail for some v3 user ... but my device is working. 

I was flashing the latest factory image and afterwards the latest v3 Image. 

i only had to do: 
ifconfig wlan0 up
and 
iw dev wlan0 scan is showing the local APs. 

i checled also the logs provided by the users once again, 
looks okay sofar ... 

I guess the location where you purchased the deivce dos'nt matter.
As then you would see at least some wifi networks. 

Strange :/ 


BR,

blueline

adrianschmutzler commented:
@apcameron I've reported my iwinfo and iw data in the forum thread:
https://forum.openwrt.org/t/support-for-tp-link-cpe210-v3-0/20208/44

But I cannot draw any conclusions from them.

adrianschmutzler commented:
I'm working on TP-Link TP-WR841N v10-v12 at the moment, which are also qca9533. I'm curious what WiFi does on them.

bluelineXY commented:
I will try to find a cpe210v2 in our local Freifunk community - i need the HW for futher testing.

apcameron commented:
What version of the qca9533 was the TP-Link TP-WR841N v1
The CPE210 v2 is version 2
SoC: Qualcomm Atheros QCA9533 ver 2 rev 0
